compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/inlineclasses/MemoizedInlineClassReplacements.kt

```diff
@@ -176,33 +176,41 @@ class MemoizedInlineClassReplacements {
         replacementOrigin: IrDeclarationOrigin,
         noFakeOverride: Boolean = false,
         body: IrFunctionImpl.() -> Unit
-    ) =
-        buildFunWithDescriptorForInlining(function.descriptor) {
-            updateFrom(function)
-            origin = if (function.origin == IrDeclarationOrigin.GENERATED_INLINE_CLASS_MEMBER) {
-                JvmLoweredDeclarationOrigin.INLINE_CLASS_GENERATED_IMPL_METHOD
-            } else {
-                replacementOrigin
-            }
-            if (noFakeOverride) {
-                isFakeOverride = false
-            }
-            name = mangledNameFor(function)
-            returnType = function.returnType
-        }.apply {
-            parent = function.parent
-            annotations += function.annotations
-            copyTypeParameters(function.allTypeParameters)
-            metadata = function.metadata
-            function.safeAs<IrFunctionBase<*>>()?.metadata = null
-
-            if (function is IrSimpleFunction) {
-                correspondingPropertySymbol = function.correspondingPropertySymbol
-                overriddenSymbols = function.overriddenSymbols.map {
-                    getReplacementFunction(it.owner)?.symbol ?: it
+    ) = buildFunWithDescriptorForInlining(function.descriptor) {
+        updateFrom(function)
+        origin = if (function.origin == IrDeclarationOrigin.GENERATED_INLINE_CLASS_MEMBER) {
+            JvmLoweredDeclarationOrigin.INLINE_CLASS_GENERATED_IMPL_METHOD
+        } else {
+            replacementOrigin
+        }
+        if (noFakeOverride) {
+            isFakeOverride = false
+        }
+        name = mangledNameFor(function)
+        returnType = function.returnType
+    }.apply {
+        parent = function.parent
+        annotations += function.annotations
+        copyTypeParameters(function.allTypeParameters)
+        metadata = function.metadata
+        function.safeAs<IrFunctionBase<*>>()?.metadata = null
+
+        if (function is IrSimpleFunction) {
+            val propertySymbol = function.correspondingPropertySymbol
+            correspondingPropertySymbol = propertySymbol
+            if (propertySymbol != null) {
+                when (function) {
+                    propertySymbol.owner.getter -> propertySymbol.owner.getter = this
```

Comment:
Suppose we have a property accessor declaration with inline class types in the signature in `decl.kt`, and its usage in `use.kt`, and we're lowering `use.kt` first. Looks like we'll reach this place when we encounter the call site, change the `IrProperty`'s accessor and this would probably mess up the subsequent run of lowerings on `decl.kt`, in particular `JvmPropertiesLowering`, because we'll observe the already modified declaration there. Or at least it would make the lowerings' behavior non-trivially dependent on the file order.

I wonder if we can do something about it? In general it seems that modifying any declaration going through the symbol link (`correspondingPropertySymbol`, call's `symbol`, `overriddenSymbols` etc.) is risky because it can be declared in another file.

Reply:
Probably because of this issue, compilation of kotlin-stdlib with JVM IR fails:
```
  Caused by: java.lang.RuntimeException: Exception while generating code for:
  FUN STATIC_INLINE_CLASS_REPLACEMENT name:getSize-impl visibility:public modality:OPEN <> (arg0:kotlin.UByteArray) returnType:kotlin.Int
    correspondingProperty: PROPERTY name:size visibility:public modality:OPEN [val]
    overridden:
      public abstract fun <get-size> (): kotlin.Int declared in kotlin.collections.Collection
    VALUE_PARAMETER MOVED_DISPATCH_RECEIVER name:arg0 index:0 type:kotlin.UByteArray
    at org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.generate(FunctionCodegen.kt:50)
    at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generateMethodNode(ClassCodegen.kt:261)
    at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generateMethod(ClassCodegen.kt:278)
    at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generate(ClassCodegen.kt:119)
    at org.jetbrains.kotlin.backend.jvm.JvmBackendFacade.doGenerateFilesInternal$backend_jvm(JvmBackendFacade.kt:120)
    ... 37 more
  Caused by: java.lang.NullPointerException
    at org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen.generate(ExpressionCodegen.kt:201)
    at org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.doGenerate(FunctionCodegen.kt:105)
    at org.jetbrains.kotlin.backend.jvm.codegen.FunctionCodegen.generate(FunctionCodegen.kt:48)
    ... 41 more
```

Reply:
You're right, that was a terrible idea. I've changed the code to create a new property instead and link it up properly with the replaced getter/setter. I've also added a test for the case you mentioned (which did indeed fail with the previous code).